matlab/polar-traversal/angular_hit.m

```diff
@@ -17,89 +19,145 @@
-        end
-    if ray_direction(2) > tol
-        % change in theta is dependent on the slope of the line
-        interval_theta = interval_theta - pi;
-        tStepTheta = - abs(current_voxel_ID_theta - interval_theta(1)/delta_theta);
-    else
-        interval_theta = interval_theta + pi;
-        tStepTheta = abs(current_voxel_ID_theta - interval_theta(1)/delta_theta);
-    end
-elseif abs(zmax(1) - zmin(1)) < tol && zmin(2) -t > tol
-    % hit the min & max boundary simultaneously. 
-    tMaxTheta = zmin(2);
-    if verbose
-        fprintf("hit origin min t\n")
-    end
-    if ray_direction(2) > tol
-        % change in theta is dependent on the slope of the line
-        interval_theta = interval_theta - pi;
-        tStepTheta = - abs(current_voxel_ID_theta - interval_theta(1)/delta_theta);
+if parallel_max == 0
+    a = perpvw_max / perpuv_max;
+    b = perpuw_max / perpuv_max;
+    if (a < 0 || a > 1) || (b < 0 || b > 1)
+        intersect_max = 0;
     else
-        interval_theta = interval_theta + pi;
-        tStepTheta = abs(current_voxel_ID_theta - interval_theta(1)/delta_theta);
+        intersect_max = 1;
+        p_max = p + b * v;
     end
 else
-    tStepTheta = 0;
-    tMaxTheta = inf;
-    if verbose
-        fprintf("no hit\n")
+    intersect_max = 0;
+end
+
+if ~approximatelyEqual(ray_direction(1),0.0,1e-12,1e-8)
+    inv_dir = 1/ray_direction(1);
+else
+    inv_dir = 1/ray_direction(2);
+end
+
+if intersect_max == 1
+    if ~approximatelyEqual(ray_direction(1),0.0,1e-12,1e-8)
+        inv_dir = 1/ray_direction(1);
+        t_max = (p_max(1) - ray_origin(1)) * inv_dir;
+    else
+        inv_dir = 1/ray_direction(2);
+        t_max = (p_max(2) - ray_origin(2)) * inv_dir;
     end
 end
 
+if intersect_min == 1
+    if ~approximatelyEqual(ray_direction(1),0.0,1e-12,1e-8)
+        inv_dir = 1/ray_direction(1);
+        t_min = (p_min(1) - ray_origin(1)) * inv_dir;
+    else
+        inv_dir = 1/ray_direction(2);
+        t_min = (p_min(2) - ray_origin(2)) * inv_dir;
+     end
+end
+
+if intersect_max == 1 && intersect_min == 0 && ...
+        t < t_max && t_max < t_end && ...
+        ~approximatelyEqual(t,t_max,1e-12,1e-8)
+    tStepTheta = 1;
+    tMaxTheta = t_max;
+elseif intersect_min == 1 && intersect_max == 0 && ...
+        t < t_min && t_min < t_end && ...
+        ~approximatelyEqual(t,t_min,1e-12,1e-8)
+    tStepTheta = -1;
+    tMaxTheta = t_min;
+elseif intersect_min == 1 && intersect_max == 1
+    % hit the min & max boundary simultaneously.
+    if  approximatelyEqual(t_min,t_max,1e-12,1e-8) && ...
+            t < t_min && t_min < t_end && ...
+            ~approximatelyEqual(t,t_min,1e-12,1e-8)
+        tMaxTheta = t_max;
+         if verbose
+            fprintf("hit origin t\n")
+        end
+        if ~approximatelyEqual(ray_direction(2),0.0,1e-12,1e-8)
+            % change in theta is dependent on the slope of the line
+            tStepTheta = - num_angular_sections/2;
```

Comment:
Does this also work for odd number of angular sections?

Reply:
Probably not; need conditional for odd case.